portal-impl/src/com/liferay/portal/db/remover/PortalDuplicateRemover.java

```diff
@@ -0,0 +1,212 @@
+
+	@Override
+	public Map<Long, List<HashMap<String, String>>> getDuplicates(
+		String tableName, String indexesSQL) {
+
+		Map<String[], String> indexesDuplicatesMap = new LinkedHashMap<>();
+
+		Map<Long, List<HashMap<String, String>>> duplicatesMap =
+			new LinkedHashMap<>();
+
+		List<String> indexesColumnsList = getIndexesColumnsList(indexesSQL);
+
+		for (String indexColumns : indexesColumnsList) {
+			StringBundler sb = new StringBundler(7);
+
+			sb.append("SELECT ");
+			sb.append(indexColumns);
+			sb.append(" FROM");
+			sb.append(tableName);
+			sb.append(" GROUP BY ");
+			sb.append(indexColumns);
+			sb.append(" HAVING COUNT(*) > 1;");
+			String sql = sb.toString();
+
+			try (Connection connection = DataAccess.getConnection();
+				PreparedStatement preparedStatement1 =
+					connection.prepareStatement(sql);
+				ResultSet resultSet = preparedStatement1.executeQuery()) {
+
+				ResultSetMetaData metaData = resultSet.getMetaData();
+
+				int columnCount = metaData.getColumnCount();
+
+				String[] columnResults = new String[columnCount];
+
+				while (resultSet.next()) {
+					for (int i = 1; i <= columnCount; i++) {
+						String value = resultSet.getString(i);
+
+						columnResults[i - 1] = value;
+					}
+
+					indexesDuplicatesMap.put(columnResults, indexColumns);
+				}
+			}
+			catch (Exception e) {
+				throw new RuntimeException(e);
+			}
+		}
+
+		Long duplicateGroupNumber = 1L;
+
+		for (Map.Entry<String[], String> entry :
+				indexesDuplicatesMap.entrySet()) {
+
+			String[] results = entry.getKey();
+
+			String[] columns = entry.getValue(
+			).split(
+				", "
+			);
+
+			StringBundler sb = new StringBundler();
+
+			sb.append("SELECT * FROM");
+
+			sb.append(tableName);
+
+			sb.append(" WHERE ");
+
+			for (int i = 0; i < columns.length; i++) {
+				sb.append(columns[i]);
+				sb.append(" = ");
+				sb.append(results[i]);
+
+				if (i < (columns.length - 1)) {
+					sb.append(" AND ");
+				}
+			}
+
+			String sql = sb.toString();
+
+			List<HashMap<String, String>> queryResult = new ArrayList<>();
+
+			try (Connection connection = DataAccess.getConnection();
+				PreparedStatement preparedStatement1 =
+					connection.prepareStatement(sql);
+				ResultSet resultSet = preparedStatement1.executeQuery()) {
+
+				ResultSetMetaData metaData = resultSet.getMetaData();
+
+				int columnCount = metaData.getColumnCount();
+
+				String[] columnNames = new String[columnCount];
+
+				for (int i = 1; i <= columnCount; i++) {
+					String columnName = metaData.getColumnName(i);
+
+					columnNames[i - 1] = columnName;
+				}
+
+				while (resultSet.next()) {
+					HashMap<String, String> queryMap = new HashMap<>();
+
+					for (int i = 0; i < columnCount; i++) {
+						String value = resultSet.getString(columnNames[i]);
+
+						queryMap.put(columnNames[i], value);
+					}
+
+					queryResult.add(queryMap);
+
+					duplicatesMap.put(duplicateGroupNumber, queryResult);
```

Comment:
Instead of duplicate group number,  collect the index used to indicate why the entry was removed